### PR TITLE
Dynamic rating and default background.

### PR DESCRIPTION
**Pull request by JasonLG1979**
Dynamic rating and default background/cover based on window theme's color and icon theme.

### CODE REVIEW

**Comment by TingPing** [comment]
I'm going to save this for the next release. Is there anything else you want finished for 1.2?

**Comment by JasonLG1979** [comment]
Not really. The async album art stuff needs a lot a work.

**Comment by JasonLG1979** [comment]
I'd say go for 1.2 if you think it's ready.

**Comment by JasonLG1979** [comment]
I still like this for the next release.

File: pithos/pithos.py [hunk 1]
```diff
@@ -52,6 +52,17 @@
...
 TEXT_X_PADDING = 12
 
+RATING_BG_SVG = '''
+<svg height="20" width="20">
+<g transform="translate(0,-1032.3622)">
+<path d="m 12,1032.3622 a 12,12 0 0 0 -12,12 12,12 0 0 0 3.0742188,
+8 l 16.9257812,0 0,-16.9277 a 12,12 0 0 0 -8,-3.0723 z"
+style="fill:{bg}" /></g></svg>
+'''
+
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [W293] blank line contains whitespace

File: pithos/pithos.py [hunk 2]
```diff
@@ -71,14 +82,19 @@ def state(self):
             return Gst.State.PAUSED
         elif value == 4:
             return Gst.State.NULL
-
-
+
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [W293] blank line contains whitespace

File: pithos/pithos.py [hunk 3]
```diff
@@ -71,14 +82,19 @@ def state(self):
             return Gst.State.PAUSED
         elif value == 4:
             return Gst.State.NULL
-
-
+          
+
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [W293] blank line contains whitespace

File: pithos/pithos.py [hunk 4]
```diff
@@ -90,21 +106,96 @@ def do_set_property(self, pspec, value):
...
+                rating_icon = self.ban_icon
+
+            x = x + (rating_icon.get_width() // 2)
+            y = y + (rating_icon.get_height() // 2)
+
+            Gdk.cairo_set_source_pixbuf(ctx, rating_icon, x, y)
             ctx.paint()
 
+    def update_icons(self, style_context):
+        #Dynamically change the color of backgrounds and icons
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E265] block comment should start with '# '

File: pithos/pithos.py [hunk 5]
```diff
@@ -90,21 +106,96 @@ def do_set_property(self, pspec, value):
...
+        #to match the current theme at theme changes.
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E265] block comment should start with '# '

File: pithos/pithos.py [hunk 6]
```diff
@@ -90,21 +106,96 @@ def do_set_property(self, pspec, value):
...
+        #Attempt to look up the background and foreground colors
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E265] block comment should start with '# '

File: pithos/pithos.py [hunk 7]
```diff
@@ -90,21 +106,96 @@ def do_set_property(self, pspec, value):
...
+        #in the theme's CSS file. Otherwise if they aren't found
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E265] block comment should start with '# '

File: pithos/pithos.py [hunk 8]
```diff
@@ -90,21 +106,96 @@ def do_set_property(self, pspec, value):
...
+        #fallback to get_background_color(Deprecated since version 3.16)
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E265] block comment should start with '# '

File: pithos/pithos.py [hunk 9]
```diff
@@ -90,21 +106,96 @@ def do_set_property(self, pspec, value):
...
+        #and get_color. *Most* new themes use 'theme_bg_color' and 'theme_fg_color'.
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E265] block comment should start with '# '

File: pithos/pithos.py [hunk 10]
```diff
@@ -90,21 +106,96 @@ def do_set_property(self, pspec, value):
...
+        #Some(older) themes use 'bg_color' and 'fg_color'.(like Ubuntu light themes)
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E265] block comment should start with '# '

File: pithos/pithos.py [hunk 11]
```diff
@@ -90,21 +106,96 @@ def do_set_property(self, pspec, value):
...
+        if not fg_bool:
+            fg_bool, fg_color = style_context.lookup_color('fg_color')
+            if not fg_bool:
+                fg_color = style_context.get_color(Gtk.StateFlags.NORMAL)
+                logging.debug("Could not find 'theme_fg_color' or 'fg_color' with lookup_color.")
+
+        fg_rgb = fg_color.to_string()
+        bg_rgb = bg_color.to_string()
+
+        #Use our color values to create strings representing valid SVG's
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E265] block comment should start with '# '

File: pithos/pithos.py [hunk 12]
```diff
@@ -90,21 +106,96 @@ def do_set_property(self, pspec, value):
...
+        #for backgound and rate_bg, then load them with PixbufLoader.
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E265] block comment should start with '# '

File: pithos/pithos.py [hunk 13]
```diff
@@ -90,21 +106,96 @@ def do_set_property(self, pspec, value):
...
+            loader.write(background)
+        self.background = loader.get_pixbuf()
+
+        with contextlib.closing(GdkPixbuf.PixbufLoader()) as loader:
+            loader.write(rating_bg)
+        self.rate_bg = loader.get_pixbuf()
+
+        current_theme = Gtk.IconTheme.get_default()
+
+        #Pithos requires an icon theme with symbolic icons.
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E265] block comment should start with '# '

File: pithos/pithos.py [hunk 14]
```diff
@@ -90,21 +106,96 @@ def do_set_property(self, pspec, value):
...
+
+        #Manually color audio-x-generic-symbolic 48px icon to be used as part of the "default cover".
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E265] block comment should start with '# '

File: pithos/pithos.py [hunk 15]
```diff
@@ -90,21 +106,96 @@ def do_set_property(self, pspec, value):
...
+        info = current_theme.lookup_icon('audio-x-generic-symbolic', 48, 0)
+        self.generic_audio_icon, was_symbolic = info.load_symbolic(bg_color, bg_color, bg_color, bg_color)
+
+        #We request 24px icons because what we really want is 12px icons,
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E265] block comment should start with '# '

File: pithos/pithos.py [hunk 16]
```diff
@@ -90,21 +106,96 @@ def do_set_property(self, pspec, value):
...
+        #and they doesn't exist in many(or any?) icon themes. We then manually color
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E265] block comment should start with '# '

File: pithos/pithos.py [hunk 17]
```diff
@@ -90,21 +106,96 @@ def do_set_property(self, pspec, value):
...
+        #and scale them down to 12px.
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E265] block comment should start with '# '

File: pithos/pithos.py [hunk 18]
```diff
@@ -328,6 +419,22 @@ def init_actions(self, app):
         app.add_accelerator('<Primary>d', 'win.bookmark', None)
         action.connect('activate', self.bookmark_song)
 
+    def style_context_change_handler(self, style_context):
+        #We only care if the window or icon theme has changed.
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E265] block comment should start with '# '

File: pithos/pithos.py [hunk 19]
```diff
@@ -328,6 +419,22 @@ def init_actions(self, app):
...
+        current_theme = Gtk.Settings.get_default()
+        current_window_theme_name = current_theme.props.gtk_theme_name
+        current_icon_theme_name = current_theme.props.gtk_icon_theme_name
+        if (self.current_window_theme_name != current_window_theme_name
+            or self.current_icon_theme_name != current_icon_theme_name):
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E129] visually indented line with same indent as next logical line
* [W503] line break before binary operator

File: pithos/pithos.py [hunk 20]
```diff
@@ -328,6 +419,22 @@ def init_actions(self, app):
...
+            self.current_window_theme_name = current_window_theme_name
+            self.current_icon_theme_name = current_icon_theme_name
+            #Update rating icons and background, and generic cover icon and background.
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E265] block comment should start with '# '

File: pithos/pithos.py [hunk 21]
```diff
@@ -328,6 +419,22 @@ def init_actions(self, app):
...
+            self.render_cover_art.update_icons(style_context)
+            #If the current song is using the generic icon as the cover
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E265] block comment should start with '# '

File: pithos/pithos.py [hunk 22]
```diff
@@ -328,6 +419,22 @@ def init_actions(self, app):
...
+            #we need to send a metadata-changed signal so MPRIS updates the generic icon cover also.
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E265] block comment should start with '# '

**Comment by TingPing** [comment]
Probably should remove the old default icon too.

**Pull request merged**